sycl/include/sycl/detail/kernel_launch_helper.hpp

```diff
@@ -251,20 +254,325 @@ struct KernelWrapper<
+                FeatureT>
+  static void throwIfGraphAssociated(bool HasGraph) {
+    if (!HasGraph)
+      return;
+
+    std::string FeatureString =
+        sycl::ext::oneapi::experimental::detail::UnsupportedFeatureToString(
+            FeatureT);
+    throw sycl::exception(sycl::make_error_code(errc::invalid),
+                          "The " + FeatureString +
+                              " feature is not yet available "
+                              "for use with the SYCL Graph extension.");
+  }
+
+  static void verifyDeviceHasProgressGuarantee(
+      bool supported,
+      sycl::ext::oneapi::experimental::forward_progress_guarantee guarantee,
+      sycl::ext::oneapi::experimental::execution_scope threadScope,
+      KernelLaunchPropertiesT &prop) {
+
+    using execution_scope = sycl::ext::oneapi::experimental::execution_scope;
+    using forward_progress =
+        sycl::ext::oneapi::experimental::forward_progress_guarantee;
+
+    if (threadScope == execution_scope::work_group) {
+      if (!supported) {
+        throw sycl::exception(
+            sycl::errc::feature_not_supported,
+            "Required progress guarantee for work groups is not "
+            "supported by this device.");
+      }
+      // If we are here, the device supports the guarantee required but there is
+      // a caveat in that if the guarantee required is a concurrent guarantee,
+      // then we most likely also need to enable cooperative launch of the
+      // kernel. That is, although the device supports the required guarantee,
+      // some setup work is needed to truly make the device provide that
+      // guarantee at runtime. Otherwise, we will get the default guarantee
+      // which is weaker than concurrent. Same reasoning applies for sub_group
+      // but not for work_item.
+      // TODO: Further design work is probably needed to reflect this behavior
+      // in Unified Runtime.
+      if (guarantee == forward_progress::concurrent)
+        prop.MKernelIsCooperative = true;
+    } else if (threadScope == execution_scope::sub_group) {
+      if (!supported) {
+        throw sycl::exception(
+            sycl::errc::feature_not_supported,
+            "Required progress guarantee for sub groups is not "
+            "supported by this device.");
+      }
+      // Same reasoning as above.
+      if (guarantee == forward_progress::concurrent)
+        prop.MKernelIsCooperative = true;
+    } else { // threadScope is execution_scope::work_item otherwise undefined
+             // behavior
+      if (!supported) {
+        throw sycl::exception(
+            sycl::errc::feature_not_supported,
+            "Required progress guarantee for work items is not "
+            "supported by this device.");
+      }
+    }
+  }
+
+  /// Process runtime kernel properties.
+  ///
+  /// Stores information about kernel properties into the handler.
+  template <typename PropertiesT, typename PropertyProcessorPtrT>
```

Comment:
`PropertyProcessorPtrT` needs more documentation in the comment above. What you could also do is

```
template <typename Self>
struct PropertiesProcessor {
  void foo() {
    using self = static_cast<Self *>(this);
    self->bar();
  }
};

struct handler : public PropertiesProcessor<handler> { ... };
```

IIUC what you're trying to achieve here.